Add first Oneof prototype

aleksZubakov opened the pull request:
Seems to work. Not sure if it's fully complete, but some tests work. 

Not sure also if such implementation of OneOf is good one:
- conceptually I added two kind of `Field` now: OneOfField and OneOfPartField
- OneOfField is not used for deserialization (I tweaked
  MessageSerializer a bit)
- OneOfPartField's are not used to serialiaze, OneOfField is used for it
- For every  submessage there is separate `OneOfPartField`, and they kinda know about it's parent (or enclosing one?)
   For example submessages here are `name` and `sub_message`, parent is `test_oneof`: 
   ```python
    # message SubMessage {
    #     string id = 1;
    #     int32 b = 3;
    # }
    #
    # message SampleMessage {
    #   oneof test_oneof {
    #     string name = 4;
    #     SubMessage sub_message = 9;
    #   }
    # }

    @message
    @dataclass
    class SubMessage:
        id: str = field(1)
        b: int32 = field(3)

    @message
    @dataclass
    class SampleMessage:
        test_oneof: OneOf_ = one_of(
            name=part(str, 4),
            sub_message=part(SubMessage, 9)
        )
    ``` 
 
- I didn't implement distinct Serilazers for this Fields, but probably should

Mb will clean up later and rename some methods/variables. 

Anyway design on user side looks good to me now. Probably better to start here: function [test_one_of_field](https://github.com/eigenein/protobuf/compare/master...aleksZubakov:one-of-implementation?expand=1#diff-40fa549b2e04aff4adbd50658b9bfff04fb9750a75eda1cb415ef005818e258eR275) (hope link gonna work) in `test_serializers.py` first.

aleksZubakov commented:
Seems that checks for 3.10 don't work because of some HTTP Error

If you don't mind I will ignore DeepSource warnings for now and will fix them if needed later.

eigenein commented:
Thanks @aleksZubakov!

I've just deactivated DeepSource (it's been bothering me for long) and fixed the issue with Coveralls. So if you rebase, it _should_ become green.

I expect I'll have time to take a deeper look at the PR in the coming days, perhaps tomorrow if nothing breaks my plans.

aleksZubakov commented:
Cleaned up implementation, now it looks way better:
- Now there is only `OneOfPartField`, its methods `validate` and `dump` are written in such way that they work only if corresponding field is set in passed `OneOf_`;
- `OneOfField` is droped, it's not needed anymore;
- No need in separate serializer/deserializer, `OneOf_` is all about the sugar;
- `OneOf_` as other parts seems to be fully complete;
- Added more tests, but any new are welcome.

Think I'm done here for a while, haven't written examples for readme yet, but mb will have.

eigenein commented on pure_protobuf/dataclasses_.py:
```diff
@@ -102,12 +83,22 @@ def loads(cls: Type[TMessage], bytes_: bytes) -> TMessage:
+
+
+def field(number: int, *args, packed=True, **kwargs) -> Any:
+    return _field(number, *args, packed=packed, **kwargs)
```
Could you elaborate on why this «proxy»-function is needed?

aleksZubakov replied:
I just needed to pass `isoneof` attribute to `metadata` (it can be dropped but better to use it than call `isinstance` function to find `OneOf_` fields later) and I intended here to reuse code of previous function `field` and not rewrite field init logic. 

I could pass `isoneof` into `field` itself but should we expose `isoneof` arg for usual fields in usual classes? I mean addition to list of args to previous func like:
```python 
def field(number: int, *args, packed=True, isoneof=False, **kwargs) -> Any:
```
seems to be a bad idea.

So I moved code a bit and renamed. Now it's invoked from `field` and later from `one_of`.

Mb should just copy and inline code of `_field` into body of `one_of`, don't know.

eigenein replied:
Yeah, I've just overlooked the presence of `isoneof=False` which you hide for a purpose. Let's keep it like that

eigenein commented on pure_protobuf/dataclasses_.py:
```diff
@@ -117,6 +108,21 @@ def optional_field(number: int, *args, **kwargs) -> Any:
+
+
+def part(type_: Type, number: int) -> Tuple[type, int]:
+    # well, yeah
```
Why? 😄

aleksZubakov replied:
Mostly because why not :)

Tuples are small and immutable. They are enough for declaring such a message. 

And `part` is just a good constructor name for user to understand what's going on. Ofc it can be simplified to just tuple passing, but should we?

eigenein replied:
I'd just use a tuple, yet not strictly against the function.

But do you only use it in the tests? Because then I'd prefer to have it in `tests`

aleksZubakov replied:
> But do you only use it in the tests? 

I actually thought of it as of part of public API.

> yet not strictly against the function.

Ok, let's leave it as is for now.

Probably function has advantage in such case:

I changed signature a bit for case when `packed` flag needed:
```python
def part(type_: Type, number: int, *, packed: bool = False) -> Tuple[type, int, bool]:
    ...
```
so user must pass `packed` as named arg:
```python
@message
class A:
    a: OneOf_ = one_of(
        foo=part(int, 1, packed=True)
    )
```

But I do not insist here, it can be changed in every possible way I think

eigenein commented on pure_protobuf/dataclasses_.py:
```diff
@@ -117,6 +108,21 @@ def optional_field(number: int, *args, **kwargs) -> Any:
+    )
+
+    default = OneOf_(scheme_)
+    return _field(-1, isoneof=True, default=default)
```
Does the `-1` get handled somewhere? I must be overlooking it

aleksZubakov replied:
Kinda yes. The trick here is that from `OneOf_` there will be created fields in   `__protobuf_fields__` only for its parts, and `OneOf_` is never a part of a message, it's only sugar.

For example for `A`:
```python
@message
@dataclass
class A:
    msg: OneOf_ = one_of(
        a=part(1, int32),
		b=part(2, str)
    )
```

`__protobuf_fields__` gonna look like that:

```python
__protobuf_fields__ = {
    1: OneOfPartField(1, name="msg", origin=Int32Field(name='a', ...), ...),
    2: OneOfPartField(2, name="msg", origin=StrField(name='b', ...), ...),
}
```
there won't be any field for `OneOf_`i.e. for `msg` itself, only for it's parts

eigenein replied:
Hmm, yeah I see. I don't really like passing around a fake value which is never used. It suggests that we should split `_field` function. Maybe later (tm)

eigenein commented on tests/test_oneof.py:
```diff
@@ -0,0 +1,213 @@
+    with raises(TypeError):
+        one_of_value = OneOf_((OneOfPartInfo('complex', ComplexObj, 1), ))
+        one_of_value.complex = ComplexObj(a=5, b=1.)
+        hash(one_of_value)
```
Could you keep only the `hash(…)` call here and move the rest out of `with raises(…)`?

aleksZubakov replied:
Yep, good idea

eigenein commented on tests/test_oneof.py:
```diff
@@ -0,0 +1,213 @@
+        OneOfPartInfo('foo', int, 10),
+        OneOfPartInfo('f1', ComplexObj, 10),
+        OneOfPartInfo('f2', List[int], 9)
+    )
```
Could you move these to `pytest.mark.parametrize`?

aleksZubakov replied:
This are not parameters for separate measurements/runs, it is the same scheme of one `OneOf_` obj, and 

And same stuff here:
```python
oneof_msg.foo = expected  # every set actually overrides
```

This could be changed to `getattr` call, but I decided to hardcode.

If needed, it can be dropped.

eigenein replied:
Sorry it's me, I've overlooked it

eigenein commented on tests/test_oneof.py:
```diff
@@ -0,0 +1,213 @@
+    assert _r_val == expected_val
+
+
+def test_class_sets_one_set():
```
Could you make this a parametrized test too?

aleksZubakov replied:
I would be happy, but intended here to have fair attribute get from object. Technically it's same as calling `getattr`, but is it?

```python
obj = Example(usual_value=5, another_one='111')

# foo
expected, expected_name = 5, 'foo'
obj.oneof_msg.foo = expected  # every set actually overrides previous value
class_assertions(expected_name, expected, obj.oneof_msg)
```

eigenein replied:
I've mistaken, thanks!

eigenein commented on pure_protobuf/dataclasses_.py:
```diff
@@ -144,9 +159,30 @@ def message(cls: Type[T]) -> Type[TMessage]:
+    Returns the corresponding ``Field`` instance.
+    """
+    child_fields = (
+        # TODO: what to do with packed?
```
What's exactly the issue here? What'd happened if you just bypass the `packed` flag?

aleksZubakov replied:
> What's exactly the issue here?

There is no actually, just haven't known what to do and should it be exposed as part of api or not.

You mean pass it right into `part`?

eigenein replied:
Yep. I guess the `packed` flag may be safely passed from the declaration through the part info to the `make_field` call

eigenein commented on pure_protobuf/oneof.py:
```diff
@@ -0,0 +1,91 @@
+    return obj.__parts__
+
+
+def internals(self) -> Tuple[Any, Any, Any]:
```
Wouldn't it be more readable to just use the attributes directly instead of wrapping it into a function returning `Any`?

aleksZubakov replied:
It's used only in this file for `__eq__`, `__hash__`, `__repr__` implementation and likely should be named `_internals`. Not supposed to be used somewhere else.

eigenein reviewed (commented):
Thanks for taking it!

I must admit though that I barely understand what's going on 🙈  But that's not to blame – maybe you could add some comments explaining the data flows?

aleksZubakov commented:
> maybe you could add some comments explaining the data flows?

yeah, gonna try

aleksZubakov commented:
Thanks for a response and sorry for such amount of probably weird things :) 

Will write it all as docstrings later I think, but texted something here.

## What's going on here?

### OneOf_

`OneOf_` is a simple beast. It knows the scheme, as a consequence it knows all
its fields names and actually looks like an object with attributes corresponding
to scheme.

Minimal code to declare `OneOf_` with given scheme. 
```python
from pure_protobuf.oneof import OneOf_, OneOfPartInfo
from pure_protobuf.types import int32

oneof_ = OneOf_(tuple(
    OneOfPartInfo('a', int32, 1),
    OneOfPartInfo('b', str, 2)
))

```
Scheme as you can see is just a list of parts where every part contains info 
about field of `oneof`.

`__init__` of `OneOf_` is ugly, `super().__setattr__` calls are must

`OneOf_` works through implementing `__getattr__`, `__setattr__`, `__delattr__`.
There is outside helper to implement this methods: `_name_in_attrs_check`.

Also `OneOf_` implements next dunder methods:

- `__eq__`, `__hash__` -- work only if value of field can be compared/hashable
- `__repr__` -- to debug state

And property:
- `which_one_of: Optional[str]` -- if some field of `OneOf_` value of 
`which_one_of` equals to name of this field.

And outside public function:
- `scheme(obj: OneOf_) -> Tuple[OneOfPartInfo, ...]` -- to get scheme from 
existing `OneOf_`

It's supposed that there is one `OneOf_` object for message.

### OneOfPartField

`OneOfPartField` is special implementation of `Field`.

Like all other implementations it has 

- `merge`,
- `load`,
- `dump`,
- `validate`

methods.

The trick here is that from `OneOf_` there will be created many `OneOfPartField`
instances in `__protobuf_fields__` -- one for part of a message. 

`OneOf_` itself doesn't need it's own proto\_field and serializer, it's only 
sugar.

For example for `A`:
```python
@message
@dataclass
class A:
    msg: OneOf_ = one_of(
        a=part(1, int32),
		b=part(2, str)
    )
```

`__protobuf_fields__` gonna look like that:

```python
__protobuf_fields__ = {
    1: OneOfPartField(1, name="msg", origin=Int32Field(name='a', ...), ...),
    2: OneOfPartField(2, name="msg", origin=StrField(name='b', ...), ...), # I think there is no such thing as `StrField`
}
```
`validate` and `dump` also have to be tricky and work in such way that they 
will dump passed `OneOf_` value only if currently set attribute of `OneOf_`, 
i.e. the name of this attribute is equal to original field name.
See `@OneOfPartField.do_if_current_set` decorator for details.

`load` implemented in such a way currently that it sets loaded value to 
attr of saved `OneOf_`, but this must be changed, looks like some stupid check 
in there gonna break everything pretty soon.

I think new instance should be created there.

### dataclasses_ interface

- `one_of(**parts: Tuple[type, int]) -> OneOf_:` -- entry point for creating 
`OneOf_` instances. Forces user to give names to parts.

- `part` -- function to declare part of oneof message

- `make_one_of_field(field_: OneOf_, name: str) -> Dict[int, OneOfPartField]` --
function to create fields for all parts of oneof message. So it return map
num of field -> real field itself.

eigenein reviewed (approved):
I think we could merge it, unless you'd like to improve something further.

aleksZubakov commented:
> could you squash the commits into one?

Yep, created separate branch PR to not push force and mess up comments here: [link](https://github.com/eigenein/protobuf/pull/95)
